stopcovid/status/reminder_triggers.py

```diff
@@ -0,0 +1,103 @@
+import uuid
+from dataclasses import dataclass
+from typing import List
+
+from sqlalchemy import (
+    Table,
+    MetaData,
+    Column,
+    String,
+    select,
+    exists,
+    Index,
+    and_,
+    insert,
+)
+from sqlalchemy.dialects.postgresql import UUID
+from sqlalchemy.schema import UniqueConstraint
+from sqlalchemy.exc import DatabaseError, IntegrityError
+
+from . import db
+
+metadata = MetaData()
+reminder_triggers = Table(
+    "reminder_triggers",
+    metadata,
+    Column("id", UUID, primary_key=True),
+    Column("drill_instance_id", UUID, nullable=False, index=True),
+    Column("prompt_slug", String, nullable=False),
+    # Do not remind for same prompt on drill twice
+    UniqueConstraint("drill_instance_id", "prompt_slug"),
+)
+
+
+Index(
+    "drill_instance_id_prompt_slug_index", "drill_instance_id", "prompt_slug",
+)
+
+
+@dataclass
+class ReminderTrigger:
+    id: uuid.UUID
+    drill_instance_id: uuid.UUID
+    prompt_slug: str
+
+    def serialize(self):
+        return {
+            "id": str(self.id),
+            "drill_instance_id": str(self.drill_instance_id),
+            "prompt_slug": self.prompt_slug,
+        }
+
+
+class ReminderTriggerRepository:
+    def __init__(self, engine_factory=db.get_sqlalchemy_engine):
+        self.engine_factory = engine_factory
+        self.engine = engine_factory()
+
+    def _deserialize(self, row):
+        return ReminderTrigger(
+            id=uuid.UUID(row["id"]),
+            drill_instance_id=uuid.UUID(row["drill_instance_id"]),
+            prompt_slug=row["prompt_slug"],
+        )
+
+    def save_reminder_triggers(self, values: List[ReminderTrigger]):
+        with self.engine.connect() as connection:
+            with connection.begin():
+                for value in values:
+                    stmt = insert(reminder_triggers).values(
+                        id=str(value.id),
+                        drill_instance_id=str(value.drill_instance_id),
+                        prompt_slug=value.prompt_slug,
+                    )
+                    try:
+                        self.engine.execute(stmt)
+                    except IntegrityError:
+                        pass
```

Comment:
when would this happen?

Reply:
This is me trying to gracefully handle retries. I dont want the stream to get blocked if we try to rewrite something that already exists

Reply:
cool. 🚢 though I'd log a warning here

(we should turn on logging :) )